grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,269 @@
+            case SERVICE_UNKNOWN:
+                return SERVICE_UNKNOWN_RESPONSE;
+            default:
+                throw new IllegalArgumentException("Invalid status:" + status);
+        }
+    }
+
+    private final SettableHealthChecker serverHealth;
+    private final Set<ListenableHealthChecker> serverHealthCheckers;
+    private final Map<String, ListenableHealthChecker> grpcServiceHealthCheckers;
+    private final Multimap<String, StreamObserver<HealthCheckResponse>> watchers = HashMultimap.create();
+    @Nullable
+    private Server server;
+
+    GrpcHealthCheckService(
+            Set<ListenableHealthChecker> serverHealthCheckers,
+            Map<String, ListenableHealthChecker> grpcServiceHealthCheckers,
+            List<HealthCheckUpdateListener> updateListeners
+    ) {
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(updateListeners);
+        }
+        this.serverHealthCheckers = ImmutableSet.<ListenableHealthChecker>builder()
+                                                .add(serverHealth)
+                                                .addAll(serverHealthCheckers)
+                                                .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        final Consumer<String> healthCheckUpdateListener = watcherHealthUpdater();
+        serverHealthCheckers.forEach(
+                lhc -> lhc.addListener(healthChecker -> healthCheckUpdateListener.accept(EMPTY_SERVICE)));
+        grpcServiceHealthCheckers.forEach((serviceName, lhc) -> {
+            lhc.addListener(healthChecker -> healthCheckUpdateListener.accept(serviceName));
+        });
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
+            responseObserver.onError(Status.NOT_FOUND
+                                             .withDescription(String.format(
+                                                     "The service name(%s) is not registered in this service",
+                                                     service))
+                                             .asRuntimeException());
+            return;
+        }
+        responseObserver.onNext(getHealthCheckResponse(status));
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        ServiceRequestContext.current().clearRequestTimeout();
+        final String service = request.getService() == null ? EMPTY_SERVICE : request.getService();
+        synchronized (watchers) {
+            final ServingStatus status = checkServingStatus(service);
```

Comment:
Should we send an error with `Status.NOT_FOUND` and don't put it into the `watchers` if the `status` is `SERVICE_UNKNOWN`?

Reply:
It seems to make sense.

Reply:
The implementation which is described above does not comply with the following document.

https://github.com/grpc/grpc-proto/blob/42311feab70f0bf1dc17cfd6629a80a9b8198a01/grpc/health/v1/health.proto#L52-L56

 However, since we can't add services dynamically, A status won't change from `SERVICE_UNKNOWN`.

Reply:
> since we can't add services dynamically, A status won't change from SERVICE_UNKNOWN.

That's actually what I thought. 😄 
Because we don't add a service dynamically, I thought we could just terminate the request.
However, as the document says, which it's valid let's keep it as is. 😄